### ISSUE TITLE
Create pedantic/strict mode

### ISSUE THREAD

**Issue by certik** [lfortran/lfortran#363]
*Original issue*: https://gitlab.com/lfortran/lfortran/-/issues/450

In this mode, LFortran would provide warnings (or even errors) for code that is valid, but we determined as a community that we would not like using it.

Examples:

* more restrictive intents: it can force `intent(inout)` to `intent(in)` or `intent(out)` where appropriate; and require `intent` to be present for all arguments
* apply `pure` and `simple` where appropriate
* implicit typing, implied save and other such features would not be allowed
* "1.3" and "1.3d0" would not be allowed, you have to specify the kind such as "1.3_dp" or "1.3_sp"
* `endif`, `enddo` etc. would not be allowed, you have to use `end if`, `end do`, ...
* `real(8)` and `real*8` would not be allowed, you have to use `real(dp)` and such
* standardize how to define return type of a function (see [here](https://fortran-lang.discourse.group/t/closing-if-loops-by-endif-vs-end-if/1651/7))
* not allow routines without explicit interfaces (i.e., all routines must be in modules or have an interface specified).
* When a variable is used that is not initialized, it will give a compiler error (obviously we have to ensure no false positives)
* ...

We can add any number of rules. The idea is that we determine as a community what kind of modern Fortran we would like to be writing and then use the compiler to enforce the rules. It can also be configurable.

This mode will be possible to enable / disable with a compiler option. First this mode would be off by default, but as people use it and if they like it, we can discuss if it should be on by default (you'll always be able to turn it off with an option).

**Comment by ivan-pi** [comment]
One warning/suggestion I think would be useful is in dealing with optional arguments. A safe style of programming dictates keyword arguments should be used with optional dummy variable:

```fortran
subroutine foo(a,bar,baz)
integer, intent(in) :: a
integer, intent(in), optional :: b, c
...
end subroutine
```

```fortran
integer :: bar, baz
bar = 4
baz = 5
call foo(42, baz, bar) ! oops, wrong positional order, but works fine
call foo(42, bar=bar, baz=baz) ! better
```

If the optional dummy variables have distinct types you are perhaps spared from problem. But even subtler problems can occur if you combine optional with unlimited polymorphic variables:
```fortran
subroutine foo(a,bar,baz)
integer, intent(in) :: a
class(*), intent(in), optional :: bar
integer, intent(in), optional :: baz
select type(bar)
type is (expected_type)
   ... 
end select ! omitted default section to warn upon unexpected types
end subroutine
```

```fortran
call foo(a,baz) ! baz gets passed as bar, will not error for any type
```

**Comment by certik** [comment]
Also:

* All module variables should be private by default

**Comment by p-pap** [comment]
> * more restrictive intents: it can force `intent(inout)` to `intent(in)` or `intent(out)` where appropriate

I don't know whether this is included in "where appropriate", but I think that, by default, functions should not have `intent(out)` or `intent(inout)` arguments. Just keep functions as functions in a more or less mathematical point of view. If more output is needed, let the function return a derived type or just turn it to a subroutine instead.

> * All module variables should be private by default

Just to clarify, I hope this means everything defined within a module should be `private` by default, including variables, parameters, derived types, procedures, and other `use`d modules. Everything that needs to be `public` must be explicitly stated as such. This enforces information hiding. It's basically like defining a module as `private` by default.

**Comment by certik** [comment]
@p-pap Yes, both are great ideas. Good point about enforcing functions just accepting `intent(in)` arguments. If you want `intent(out)` or `intent(inout)`, you should use a subroutine.

**Comment by aradi** [comment]
@certik As for the bullet point about dummy arguments with the `target` attribute: I don't see any way to check for proper usage without generating false positives for standard conforming cases. See [my answer on discourse](https://fortran-lang.discourse.group/t/is-pointing-to-a-target-dummy-argument-safe-after-return/5199/16?u=aradi)
